pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1075,6 +1076,45 @@ private void closeConsumerTasks() {
         }
         negativeAcksTracker.close();
         stats.getStatTimeout().ifPresent(Timeout::cancel);
+        releaseMsgIfEnabledPooledMsg();
+    }
+
+    /**
+     * If enabled pooled messages, we should release the messages after closing consumer.
+     *   1. Call "clearIncomingMessages" regardless of whether "internalPinnedExecutor" has shutdown or not.
+     *   2. Increment epoch will auto release all messages which received after close.
+     *   3. Call "clearIncomingMessages" in "internalPinnedExecutor" is used to clear messages in flight.
+     */
+    private void releaseMsgIfEnabledPooledMsg() {
+        if (!poolMessages) {
+            return;
+        }
+        // Increment epoch.
+        CONSUMER_EPOCH.incrementAndGet(this);
+        // Try clear the incoming queue in internalPinnedExecutor-thread.
+        boolean executorIsShutdown = false;
+        try {
+            if (!internalPinnedExecutor.isShutdown()) {
```

Comment:
I think we can call `internalPinnedExecutor.execute` directly...

Reply:
Exception catching is more cost than the check of `if,` so `if` is better

Reply:
Which `executorIsShutdown` value is expected when `internalPinnedExecutor.isShutdown()` is true?

Reply:
Ah, sorry. Already fixed. Thanks